### ISSUE TITLE
(#) plus (^=) has wrong behaviour

### ISSUE THREAD

**Issue by vrom911** [sebastiaanvisser/clay#206]
I have the following code:

```haskell
".row" # ("class" ^= "col") ? do
        ...
```

And it results in 
```css
.row[class^="col"] {
  ...
}
```

which doesn't work. The correct rendered one should be
```css
.row [class^="col"] {
  ...
}
```

I am using `clay-0.13.3`.

Let me know if there is a better way to do what I am trying to achieve as I am not very familiar with the library. But, meanwhile, I am not sure how to fix this behaviour :thinking: 

Thanks a lot for the library, btw :+1:

**Comment by turion** [comment]
Thanks for spotting this! I'm surprised this is a CSS problem, but my knowledge of CSS is not very great.

The operator `#` is defined here: https://github.com/sebastiaanvisser/clay/blob/master/src/Clay/Selector.hs#L67

Do you want to work on fixing this? I can help you along the way. The first step would be to go to look at the test suite, which is here: https://github.com/sebastiaanvisser/clay/tree/master/spec/Clay Then you could add a new test file `SelectorSpec.hs`, inspired by the existing files. There you add your test case. Then you can upload it as a merge request and we're going to work on it together until the issue is fixed.

**Comment by turion** [comment]
@vrom911 I'm not so sure anymore this is actually a bug. The examples here show no spaces: https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors

Can you describe more precisely (maybe with a screen shot of a website) how the rendering is incorrect when the space isn't there?

**Comment by vrom911** [comment]
@turion none of the example in the link you provided are with class. I can confirm that when it is not class it is working. However, you can check the following HTML file:

```html
<!DOCTYPE HTML>
<html>
    <head>
        <style>
            .row[class^='col'] { background-color: yellow;}
        </style>
    </head>
    <body>
        <div class="row">
            <div class="col-example"> Some text</div>
        </div>
    </body>
</html>
```

It is rendered like this:

![image](https://user-images.githubusercontent.com/8126674/84589956-5835c600-ae2a-11ea-9ef7-d1f83e722589.png)

So the specified CSS didn't work.

However, if I change just one space:

```html
<!DOCTYPE HTML>
<html>
    <head>
        <style>
            .row [class^='col'] { background-color: yellow;}
        </style>
    </head>
    <body>
        <div class="row">
            <div class="col-example"> Some text</div>
        </div>
    </body>
</html>
```

It will perfectly work:

![image](https://user-images.githubusercontent.com/8126674/84589979-81565680-ae2a-11ea-8502-01f7153d3d5a.png)

**Comment by turion** [comment]
## TL;DR

@vrom911 aaahh, thank you, that's helpful. I think the issue here is that `(#)` is _refinement_ or filter (see https://hackage.haskell.org/package/clay-0.13.3/docs/Clay.html#v:-35- , and not the descendant selector (see https://www.w3schools.com/css/css_combinators.asp). In clay, the descendant operator is `(**)` (see https://hackage.haskell.org/package/clay-0.13.3/docs/Clay.html#v:-42--42-) and it is rendered as a space, which is probably what you want, right?

## Selector combinators

For example this will work:
```
<!DOCTYPE HTML>
<html>
    <head>
        <style>
            .row[class^="col"] { background-color: yellow;}
        </style>
    </head>
    <body>
        <div class="col-example row"> Some text</div>
    </body>
</html>
```
The `div` has two classes, see https://stackoverflow.com/questions/11918491/using-two-css-classes-on-one-element. (Note that I need to match `"col"` on the beginning of the whole (space separated) list of classes, which may be a nuisance for you. In that case, maybe `*=` is a better option.)

## Descendants

But you want nested `divs`, so you have to use `**` or `deep`, because it represents the descendant selector combinator. The issue then is that `"class" ^= "col"` is not a `Selector`, but a `Refinement`, which is a bit clunky to use because `**` does not accept a `Refinement` as second argument. A workaround is this:

```
".row" ** (star # ("class" ^= "col")) ? color yellow
```

`star` matches _any_ element, so `star # myRefinement` selects all elements matching on `myRefinement`. It renders to:

```
.row [class^='col']
{
  color : #ffff00;
}
```
This is close to what you wanted I guess?

I'm not so happy with using `star` as a workaround like this, though.

## How to proceed

First of all, sorry that I led you in the wrong direction. It seems like trying to fix the behaviour of `#` wasn't what we should have done in the first place.

Instead, probably the documentation needs to be improved. There is a gap between Clay terms such as "deep", "refinement" and so on that don't match with the standard CSS terminology ("descendants", "attribute selectors"), and this gap is jointly responsible. Do you have a recommendation on how the docs could be improved such that it would have helped you towards a working solution?

Additionally, we could add a function `selectAll :: Refinement -> Selector` that converts a `Refinement` into a `Selector` using the `star` trick from above. I'm not sure about the name `selectAll` though, and will accept other suggestions.

**Comment by vrom911** [comment]
Thanks a lot for taking your time investigating the issue! This is really helpful, and I appreciate that :1st_place_medal: 

The steps you suggest to do are reasonable, I will try to apply that in my project :slightly_smiling_face: 

What about proceeding with the issue, I guess it is okay to close this one. At least if anybody would experience the same problem they would be able to find your brilliant explanation :+1:

**Comment by turion** [comment]
@vrom911 you're welcome :) feel free to start a new MR, yes. And also feel free to include the other unrelated improvements you introduced in #207!

**Comment by ddssff** [comment]
Seems important though.
